edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/util/PrometheusMetricsParser.cs

```diff
@@ -13,51 +13,135 @@ namespace Microsoft.Azure.Devices.Edge.Agent.Diagnostics.Util
+        private static readonly string Label = $"{LabelName}{Whitespace}*={Whitespace}*\"{LabelValue}\"{Whitespace}*";
         private static readonly Regex PrometheusSchemaRegex = new Regex(
-            @"^(?<metricname>[^#\{\}]+)" + // metric name. The prometheus spec says metric names should match the regex [a-zA-Z_:][a-zA-Z0-9_:]*, but it doesn't hurt to be more permissive here
-            @"(\{(?:(?:(?<tagname>[^="",]+)=(?:\""(?<tagvalue>(?:\\.|[^\\""])*)\"")" + // first tag (name="value" pair)
-            @"(?:,(?<tagname>[^="",]+)=(?:\""(?<tagvalue>(?:\\.|[^\\""])*)\""))*))?\})?\s" + // all following tags
-            @"(?<metricvalue>[\d\.eE+-]+|NaN)$", // metric value (digits and .eE+-)
+            @"^" +
+                MetricName +
+                $"{Whitespace}*" +
+                @"(?:" +
+                    @"\{" +
+                    $"{Whitespace}*" +
+                    @"(?:" +
+                        Label +
+                        @"(?:" +
+                            @"," +
+                            $"{Whitespace}*" +
+                            Label +
+                        @")*" +
+                        @"(?:" +
+                            @"," +
+                            $"{Whitespace}*" +
+                        @")?" +
+                    @")?" +
+                    @"\}" +
+                    $"{Whitespace}*" +
+                @")?" +
+                MetricValue +
+                @"(?:" +
+                    $"{Whitespace}+" +
+                    MetricTimestamp +
+                @")?" +
+            @"$",
             RegexOptions.Compiled);
 
+        private static readonly DateTime MetricTimestampEpoch = DateTime.UnixEpoch;
+
         public static IEnumerable<Metric> ParseMetrics(DateTime timeGeneratedUtc, string prometheusMessage)
         {
             Preconditions.CheckArgument(timeGeneratedUtc.Kind == DateTimeKind.Utc, $"Metric {nameof(timeGeneratedUtc)} parameter only supports in UTC.");
             using (StringReader sr = new StringReader(prometheusMessage))
             {
                 string line;
-                while ((line = sr.ReadLine()?.Trim()) != null)
+                while ((line = sr.ReadLine()) != null)
                 {
                     if (line.Length == 0 || line[0] == '#')
                     {
                         continue;
                     }
 
                     Match match = PrometheusSchemaRegex.Match(line);
-                    if (match.Success)
+                    if (!match.Success)
+                    {
+                        continue;
+                    }
+
+                    double metricValue;
+                    if (!double.TryParse(match.Groups["metricvalue"]?.Value, out metricValue))
                     {
-                        double metricValue;
-                        if (!double.TryParse(match.Groups["metricvalue"]?.Value, out metricValue))
+                        continue;
```

Comment:
This will silently over a matched metric if the value cannot be parsed to a double. Should we log a warning/error in this case?

Reply:
That is existing code.

https://github.com/Azure/iotedge/blob/9b4666fa5912692405be7150ef127170970d1aab~2/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Diagnostics/util/PrometheusMetricsParser.cs

It looks new because I reduced the indentation. See the diff with whitespace changes ignored: https://github.com/Azure/iotedge/pull/4375/files?w=1 )